### PR TITLE
feat(idempotency): handle lambda timeout scenarios for INPROGRESS records

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #1038

## Summary

Expire in-progress invocations caused by Lambda time outs. 

### Changes

> Please provide a summary of what's being changed

During an invocation, we try to figure out how much time the Lambda invocation has before it expires, and save it on the idempotency record together will all the other fields.

When a new invocation with the same key arrives, we check against this timestamp to see if there's any `INPROGRESS` invocation past the timeout timestamp. If there is, we allow the invocation to be retried.

We also moved the squence diagrams of the idempotency module in documentation to [mermaid.js](https://mermaid-js.github.io/mermaid/#/).

#### Caveats

- Only works out of the box for the `@idempotent` decorator, because right now it's the only way to access the Lambda context to get the remaining time left. I've added a `register_lambda_context` to the `IdempotencyConfig` class as a workaround

carbon (2)

### User experience

> Please share what the user experience looks like before and after this change

Before: if a Lambda invocation times out, the invocation will get stuck until the idempotency record expires (default = 1 hour).

After: if a Lambda invocation times out, the invocation will be tried again on the next retry.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/utilities/idempotency.md](https://github.com/rubenfonseca/aws-lambda-powertools-python/blob/feat/expire-inprogress/docs/utilities/idempotency.md)

### CODE REVIEW

File: aws_lambda_powertools/utilities/idempotency/config.py [hunk 1]
```diff
@@ -41,3 +43,7 @@ def __init__(
         self.use_local_cache = use_local_cache
         self.local_cache_max_items = local_cache_max_items
         self.hash_function = hash_function
+        self.lambda_context: Optional[LambdaContext] = None
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/config.py]: missing doc string. 

Future note: We should automate this with a docstring linter https://github.com/terrencepreilly/darglint

File: tests/functional/idempotency/conftest.py [hunk 2]
```diff
@@ -32,14 +31,17 @@ def lambda_apigw_event():
...
+            self.function_name = "test-func"
+            self.memory_limit_in_mb = 128
+            self.invoked_function_arn = "arn:aws:lambda:eu-west-1:809313241234:function:test-func"
+            self.aws_request_id = "52fdfc07-2182-154f-163f-5f0f9a621d72"
+
+        def get_remaining_time_in_millis(self) -> int:
+            return 1000
 
-    return namedtuple("LambdaContext", lambda_context.keys())(*lambda_context.values())
+    return LambdaContext()
```
Comment (heitorlessa) [inline tests/functional/idempotency/conftest.py:44]: Nice, must needed cleanup!

**Review by heitorlessa — CHANGES_REQUESTED** [review]
early comments before I resume reading it after lunch. At a quick glance, the pieces that would benefit the most now are: 

* **1/** Comments on the formula you're using since I saw Lambda context giving us milliseconds, then epoch (sensible for comparison), then type coercion to string (?) - sounds like we could reuse the milliseconds instead of wall clock time. Also, `datetime.datetime.now()` uses timezone offset too, instead of a safe non-backward timestamp like [time.monotonic](https://www.geeksforgeeks.org/python-time-monotonic-method/)
* **2/** I think the Update Expression would benefit to having the new one as a separate variable, then use f-string to combine them and make it more readable (which conditional is for which use case)

File: aws_lambda_powertools/utilities/idempotency/config.py [hunk 3]
```diff
@@ -41,3 +46,7 @@ def __init__(
         self.use_local_cache = use_local_cache
         self.local_cache_max_items = local_cache_max_items
         self.hash_function = hash_function
+        self.lambda_context: Optional[LambdaContext] = lambda_context
+
+    def register_lambda_context(self, lambda_context: LambdaContext):
+        self.lambda_context = lambda_context
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/config.py]:
```suggestion
    def register_lambda_context(self, lambda_context: LambdaContext):
        """Include Lambda context to calculate remaining time before it times out"""
        self.lambda_context = lambda_context
```

Docstring for public method. There's gotta be a better wording but can't think of one right now

File: aws_lambda_powertools/utilities/idempotency/base.py [hunk 4]
```diff
@@ -167,6 +190,13 @@ def _handle_for_status(self, data_record: DataRecord) -> Optional[Dict[Any, Any]
             raise IdempotencyInconsistentStateError("save_inprogress and get_record return inconsistent results.")
 
         if data_record.status == STATUS_CONSTANTS["INPROGRESS"]:
+            if data_record.in_progress_expiry_timestamp is not None and data_record.in_progress_expiry_timestamp < int(
+                datetime.datetime.now().timestamp()
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/base.py]: * 1000 for ms? tests for that too as it should've failed (as per your own word :D)

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 5]
```diff
@@ -344,6 +350,14 @@ def save_inprogress(self, data: Dict[str, Any]) -> None:
             payload_hash=self._get_hashed_payload(data=data),
         )
 
+        if remaining_time_in_millis:
+            now = datetime.datetime.now()
+            period = datetime.timedelta(milliseconds=remaining_time_in_millis)
+
+            data_record.in_progress_expiry_timestamp = int((now + period).timestamp() * 1000)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py]: maybe split into a separate variable `(now + period).timestamp` so it's easier to read that you're using the adding the delta and multiplying by 1000 for ms

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 6]
```diff
@@ -344,6 +350,14 @@ def save_inprogress(self, data: Dict[str, Any]) -> None:
...
+        else:
+            warnings.warn("Expires in progress is enabled but we couldn't determine the remaining time left")
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py]: left over? Give an actionable feedback on what they can do when reading the warning (if they do)

File: docs/utilities/idempotency.md [hunk 7]
```diff
@@ -22,6 +22,7 @@ times with the same parameters**. This makes idempotent operations safe to retry
 * Ensure Lambda handler returns the same result when called with the same payload
 * Select a subset of the event as the idempotency key using JMESPath expressions
 * Set a time window in which records with the same payload should be considered duplicates
+* Expires in-progress executions when the Lambda handler times out
```
Comment (heitorlessa) [inline docs/utilities/idempotency.md]:
```suggestion
* Expires in-progress executions if the Lambda function times out halfway through
```

File: aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py [hunk 8]
```diff
@@ -153,34 +158,51 @@ def _put_record(self, data_record: DataRecord) -> None:
...
         except self.table.meta.client.exceptions.ConditionalCheckFailedException:
             logger.debug(f"Failed to put record for already existing idempotency key: {data_record.idempotency_key}")
             raise IdempotencyItemAlreadyExistsError
 
     def _update_record(self, data_record: DataRecord):
         logger.debug(f"Updating record for idempotency key: {data_record.idempotency_key}")
-        update_expression = "SET #response_data = :response_data, #expiry = :expiry, #status = :status"
+        update_expression = (
+            "SET #response_data = :response_data, #expiry = :expiry, "
+            "#status = :status, #in_progress_expiry = :in_progress_expiry"
```
Comment (jeromevdl) [inline aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py]: The record (from save_success) doesn't contain the in_progress_expiry. This won't have any effect... Also do we actually need to add this field when the status is COMPLETED ?

Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py]: Thanks! Great comment, we really don't need it in this case, so I've removed it.

**Comment by heitorlessa** [comment]
I've pushed an updated doc with the following changes:

* Reordered the wording so we give the positive news first (what exactly we do we do?)
* Give comfort to those using `@idempotent` (When do they need to care?)
* Used banners to emphasize when they need to take action (what should I do?)
* Moved sequence diagram for latter so they're informed first, then they can dig deeper into the mechanics


image

**Comment by heitorlessa** [comment]
Shortening the wording on sequence diagram to fit in the final SVG

New (pushing shortly)

image

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**